### ISSUE TITLE
Support requested with Docker Swarm on timeouts

### ISSUE THREAD

**Issue by elijahgagne** [openfaas/faas#670]
I am trying to get started with openfaas, but hitting issues after several hours of trying.

I have three nodes running Docker version 18.03.0-ce with Swarm setup. I followed the first two labs without any issues:
- https://github.com/openfaas/workshop/blob/master/lab1.md
- https://github.com/openfaas/workshop/blob/master/lab2.md

When I get to https://github.com/openfaas/workshop/blob/master/lab3.md#hello-world-in-python I get the function deployed, but cannot figure out how to use it. 

![image](https://user-images.githubusercontent.com/2213005/39272006-b0b83806-48a8-11e8-9a6f-ba33634c9cc0.png)

```
Internal Server Error
Can't reach service: hello-openfaas
```

Looking at the logs, what I see is
```
docker service logs -f --tail 100 func_gateway
```

```
func_gateway.1.y3h5rsti4d73@server1.example.com    | 2018/04/25 18:52:40 GetHeaderCode 500
func_gateway.1.y3h5rsti4d73@server1.example.com    | 2018/04/25 18:52:40 < [http://faas-swarm:8080/function/hello-openfaas] - 500 took 3.005722 seconds
```

```
docker service logs -f --tail 100 func_faas-swarm
```

```
func_faas-swarm.1.18vsjx9yqpvy@server1.example.com    | [1524682182] Forwarding request [text/plain] to: http://hello-openfaas:8080/function/hello-openfaas
func_faas-swarm.1.18vsjx9yqpvy@server1.example.com    | 2018/04/25 18:49:45 Error, can't reach service: hello-openfaas, Post http://hello-openfaas:8080/function/hello-openfaas: dial tcp 10.0.0.41:8080: i/o timeout
```

```
docker service logs -f --tail 100 hello-openfaas
```

```
hello-openfaas.1.xwf1zxo5ehws@server2.example.com    | 2018/04/25 18:35:01 Writing lock-file to: /tmp/.lock
```

I found troubleshooting details at:
- https://blog.alexellis.io/first-faas-python-function/
- https://github.com/openfaas/faas/blob/master/guide/troubleshooting.md

And I've walked through their troubleshooting recommendations, but I'm still not able to solve this. I've also entered the containers for func_gateway, func_faas-swarm, and hello-openfaas so that I can test connecting with telnet and I'm not finding any issues there (dns seems good and telnet is connecting). I've also tried deploying the function multiple times as recommended in first-faas-python-function. 

One other piece of relavant information is that I'm accessing openfaas through a load balancer. I am browsing to https://faas-dev.example.com/ui/ and my load balancer is sending this to my three Docker servers on port 21135. To get that to work, I updated docker-compose.yml to look like:

```
version: "3.3"
services:
    gateway:
        ports:
            - 21135:8080
        image: functions/gateway:0.7.3
        networks:
            - functions
        environment:
            functions_provider_url: "http://faas-swarm:8080/"
...
```

Would anyone be willing to offer suggestions on how to troubleshoot this further?

## Your Environment

* Docker version `docker version` (e.g. Docker 17.0.05 ): 18.03.0-ce

* Are you using Docker Swarm or Kubernetes (FaaS-netes)? Swarm

* Operating System and version (e.g. Linux, Windows, MacOS): Linux

**Comment by alexellis** [comment]
Derek add label: support

**Comment by alexellis** [comment]
Sorry to hear you are having difficulties. 

I wouldn't expect you to be using 0.7.3 since the master branch has 0.7.9.. have you thought about updating before going any further?

**Comment by cpitkin** [comment]
@elijahgagne I tested lab3's hello-world on a fresh swarm cluster and it worked ok for me. I would echo what Alex said the update the version of the gateway you are using and try deploying the function again. My versions are below for reference.

Gateway: 0.7.9
Faas-swarm: 0.2.5
Docker: 18.04.0-ce
Cli: 0.6.5

**Comment by alexellis** [comment]
@elijahgagne I'm not aware of any issues with running OpenFaaS functions over multiple nodes.

Here are two suggestions:

* Try the play-with-docker environment:

https://docs.openfaas.com/deployment/

You can set up a multi-node cluster there

* Check the new troubleshooting guide

The troubleshooting guide has been moved to the documentation site now - perhaps it has more context - https://docs.openfaas.com/deployment/troubleshooting/

**Comment by elijahgagne** [comment]
Thanks. I used PWD site and saw that a 3 node swarm certainly works with openfaas. Testing in my environment more, I've found that if I run things on the two nodes that are in the same datacenter, things continue to work. When I move services to the third node that is in a different datacenter, that's when things start to break. 

I'm going to close this as it's clear the issue is in my environment - probably hardware firewall related. I really appreciate your willingness to help! If I get this working across datacenters and learn something useful from that I'll come back to this post and update it.
